File: src/cogs/reminders/reminders.py
from interactions import Extension, Permissions, SlashContext, BaseUser, Embed
from cogs.reminders.db import ReminderDB
from lib.util import subcommand
from lib.config import MODERATOR_ROLES
import re
from datetime import datetime, timedelta, timezone


def parse_duration(duration_str):
    pattern = r"(?:(\d+)w)?(?:(\d+)d)?(?:(\d+)h)?(?:(\d+)m)?(?:(\d+)s)?"
    match = re.fullmatch(pattern, duration_str)
    if not match:
        return None
    weeks, days, hours, minutes, seconds = (int(x) if x else 0 for x in match.groups())
    return timedelta(
        weeks=weeks, days=days, hours=hours, minutes=minutes, seconds=seconds
    )

def get_reminder_channel(reminder):
    if reminder["silent"]:
        return "dm"
    
    return f"<#{reminder['channel_id']}>"

class Reminders(Extension):
    """Commands for creating reminders"""

    @subcommand(
        when={
            "description": "When the reminer should be triggered. Format: (1w2d3h4m5s)"
        },
        message={"description": "What message should be attached to the reminder?"},
        silent={"description": "Should the reminder be dmed?"}
    )
    async def create(self, ctx: SlashContext, when: str, message: str, silent: bool = False):
        """Create a reminder"""

        if not any(str(role.id) in MODERATOR_ROLES for role in ctx.author.roles):  # type: ignore
            await ctx.send(
                "You do not have permission to use this command.", ephemeral=True
            )
            return

        delta = parse_duration(when)
        if delta is None or delta.total_seconds() <= 0:
            await ctx.send(
                "Invalid time format. Please use something like '1w2d3h4m5s'.",
                ephemeral=True,
            )
            return

        reminder_time = datetime.now(timezone.utc) + delta

        db = ReminderDB()
        db.add_reminder(
            remind_at=reminder_time,
            message=f"Reminder for <@{ctx.author_id}>: {message}",
            channel_id=ctx.channel_id,
            author_id=ctx.author_id,
            silent=silent
        )

        await ctx.send("Reminder is queued!", ephemeral=True)

    @subcommand(
        user={
            "description": "User whose reminders you would like to view, default is all users."
        }
    )
    async def list(self, ctx: SlashContext, user: BaseUser | None = None):
        """List the scheduled reminders for all users or a particular user"""

        if not any(str(role.id) in MODERATOR_ROLES for role in ctx.author.roles):  # type: ignore
            await ctx.send(
                "You do not have permission to use this command.", ephemeral=True
            )
            return

        db = ReminderDB()

        if user is None:
            reminders = db.get_all_reminders()
        else:
            reminders = db.get_reminders_by_author_id(user.id)

        description = ""
        for reminder in reminders:
            description += f"{reminder['id']}: {get_reminder_channel(reminder)}: '{reminder['message']}' <t:{int(reminder['remind_at'].timestamp())}:R>\n"
        
        embed = Embed(
            title="Reminders" + f" for {user.display_name}" if user else "",
            description=description or "There are no reminders currently for this query"
        )
        
        await ctx.send(embed=embed, ephemeral=True)

    @subcommand(
        reminder_id={
            "description": "The ID of the reminder to delete"
        }
    )
    async def delete(self, ctx: SlashContext, reminder_id: int):
        """Delete a reminder by its ID"""
        
        if not any(str(role.id) in MODERATOR_ROLES for role in ctx.author.roles):  # type: ignore
            await ctx.send(
                "You do not have permission to use this command.", ephemeral=True
            )
            return

        db = ReminderDB()
        
        reminder_to_delete = db.get_reminder_by_id(reminder_id)
        
        if reminder_to_delete is None:
            await ctx.send(f"Reminder with ID {reminder_id} not found.", ephemeral=True)
            return
        
        if reminder_to_delete["author_id"] != ctx.author_id and not ctx.member.has_permission(Permissions.ADMINISTRATOR): # type: ignore
            await ctx.send("You can only delete your own reminders.", ephemeral=True)
            return
        
        db.remove_reminder(reminder_id)
        await ctx.send(f"Reminder {reminder_id} has been deleted successfully!", ephemeral=True)

File: server/src/commands/reminders.ts
import {
  ApplicationCommandOptionType,
  ApplicationCommandType,
  ButtonStyle,
  ComponentType,
} from "discord-api-types/v10";
import { message } from "../discord/responses.js";
import type { Reminder } from "../storage/types.js";
import type { InteractionMessageData } from "../discord/types.js";
import { logError } from "../log.js";
import {
  deferredTask,
  finish,
  hasAdministrator,
  hasAnyRole,
} from "./helpers.js";
import {
  booleanOption,
  integerOption,
  option,
  stringOption,
  subcommand,
  userId,
} from "./options.js";
import type { CommandDefinition } from "./types.js";

const DURATION = /^(?:(\d+)w)?(?:(\d+)d)?(?:(\d+)h)?(?:(\d+)m)?(?:(\d+)s)?$/;
const PAGE_SIZE = 10;
const PAGE_DESCRIPTION_LIMIT = 3_900;
const PAGE_TTL_MS = 15 * 60_000;

export function parseDuration(value: string): number | null {
  const match = DURATION.exec(value);
  if (!match) {
    return null;
  }
  const [weeks, days, hours, minutes, seconds] = match
    .slice(1)
    .map((part) => Number(part || 0));
  const duration =
    ((((weeks * 7 + days) * 24 + hours) * 60 + minutes) * 60 + seconds) * 1_000;
  return Number.isSafeInteger(duration) ? duration : null;
}

export function renderReminderPage(
  reminders: Reminder[],
  requestedPage: number,
  stateId: string,
  title?: string,
): InteractionMessageData {
  const pages: string[][] = [[]];
  for (const reminder of reminders) {
    const line = `${reminder.id}: ${reminder.silent ? "dm" : `<#${reminder.channelId}>`}: '${reminder.message}' <t:${Math.floor(reminder.remindAt / 1_000)}:R>`;
    const current = pages.at(-1)!;
    const nextLength =
      current.join("\n").length + (current.length ? 1 : 0) + line.length;
    if (
      current.length > 0 &&
      (current.length >= PAGE_SIZE || nextLength > PAGE_DESCRIPTION_LIMIT)
    ) {
      pages.push([line]);
    } else {
      current.push(line);
    }
  }
  const pageCount = pages.length;
  const page = Math.min(Math.max(0, requestedPage), pageCount - 1);
  const description = pages[page].join("\n");
  return {
    embeds: [
      {
        ...(title ? { title } : {}),
        description:
          description || "There are no reminders currently for this query",
        footer: { text: `Page ${page + 1} of ${pageCount}` },
      },
    ],
    components:
      pageCount > 1
        ? [
            {
              type: ComponentType.ActionRow,
              components: [
                {
                  type: ComponentType.Button,
                  style: ButtonStyle.Secondary,
                  label: "Previous",
                  custom_id: `reminders-page:${stateId}:${page - 1}`,
                  disabled: page === 0,
                },
                {
                  type: ComponentType.Button,
                  style: ButtonStyle.Secondary,
                  label: "Next",
                  custom_id: `reminders-page:${stateId}:${page + 1}`,
                  disabled: page === pageCount - 1,
                },
              ],
            },
          ]
        : [],
  };
}

export const remindersCommand: CommandDefinition = {
  command: {
    name: "reminders",
    type: ApplicationCommandType.ChatInput,
    description: "No Description Set",
    options: [
      {
        type: ApplicationCommandOptionType.Subcommand,
        name: "create",
        description: "Create a reminder",
        options: [
          {
            type: ApplicationCommandOptionType.String,
            name: "when",
            description:
              "When the reminer should be triggered. Format: (1w2d3h4m5s)",
            required: true,
          },
          {
            type: ApplicationCommandOptionType.String,
            name: "message",
            description: "What message should be attached to the reminder?",
            required: true,
            max_length: 1900,
          },
          {
            type: ApplicationCommandOptionType.Boolean,
            name: "silent",
            description: "Should the reminder be dmed?",
            required: false,
          },
        ],
      },
      {
        type: ApplicationCommandOptionType.Subcommand,
        name: "list",
        description:
          "List the scheduled reminders for all users or a particular user",
        options: [
          {
            type: ApplicationCommandOptionType.User,
            name: "user",
            description:
              "User whose reminders you would like to view, default is all users.",
            required: false,
          },
        ],
      },
      {
        type: ApplicationCommandOptionType.Subcommand,
        name: "delete",
        description: "Delete a reminder by its ID",
        options: [
          {
            type: ApplicationCommandOptionType.Integer,
            name: "reminder_id",
            description: "The ID of the reminder to delete",
            required: true,
          },
        ],
      },
    ],
  },
  async handle(interaction, runtime) {
    if (
      !hasAnyRole(interaction, runtime.env.MODERATOR_ROLES) &&
      !hasAdministrator(interaction)
    ) {
      return message("You do not have permission to use this command.", true);
    }

    const action = subcommand(interaction)?.name;
    const authorId = userId(interaction);
    if (action === "create") {
      const duration = parseDuration(stringOption(interaction, "when"));
      if (duration === null || duration <= 0) {
        return message(
          "Invalid time format. Please use something like '1w2d3h4m5s'.",
          true,
        );
      }
      const remindAt = Date.now() + duration;
      const reminderMessage = `Reminder for <@${authorId}>: ${stringOption(interaction, "message")}`;
      if (reminderMessage.length > 2_000) {
        return message("Reminder message is too long.", true);
      }
      return deferredTask(interaction, runtime, true, async () => {
        await runtime.storage.createReminder({
          remindAt,
          message: reminderMessage,
          channelId: interaction.channel_id ?? "",
          authorId,
          silent: booleanOption(interaction, "silent"),
        });
        await runtime.scheduler.schedule(remindAt).catch((error) =>
          logError("reminder_schedule_failed", error, {
            interaction_id: interaction.id,
            guild_id: interaction.guild_id,
            author_id: authorId,
          }),
        );
        await finish(interaction, runtime, "Reminder is queued!");
      });
    }

    if (action === "list") {
      const requestedUserId = option(interaction, "user")?.value;
      return deferredTask(interaction, runtime, true, async () => {
        const filterAuthorId =
          typeof requestedUserId === "string" ? requestedUserId : undefined;
        const reminders = await runtime.storage.listReminders(filterAuthorId);
        const selectedUser = filterAuthorId
          ? interaction.data?.resolved?.users?.[filterAuthorId]
          : undefined;
        const title = selectedUser
          ? `Reminders for ${selectedUser.global_name ?? selectedUser.username}`
          : undefined;
        const stateId = crypto.randomUUID();
        const now = Date.now();
        await runtime.storage.createInteractionState({
          id: stateId,
          kind: "reminders_page",
          ownerUserId: authorId,
          guildId: interaction.guild_id!,
          channelId: interaction.channel_id!,
          payload: {
            ...(filterAuthorId ? { filterAuthorId } : {}),
            ...(title ? { title } : {}),
          },
          createdAt: now,
          expiresAt: now + PAGE_TTL_MS,
        });
        await finish(
          interaction,
          runtime,
          renderReminderPage(reminders, 0, stateId, title),
        );
      });
    }

    const reminderId = integerOption(interaction, "reminder_id");
    return deferredTask(interaction, runtime, true, async () => {
      const reminder = await runtime.storage.getReminder(reminderId);
      if (!reminder) {
        await finish(
          interaction,
          runtime,
          `Reminder with ID ${reminderId} not found.`,
        );
        return;
      }
      if (reminder.authorId !== authorId && !hasAdministrator(interaction)) {
        await finish(
          interaction,
          runtime,
          "You can only delete your own reminders.",
        );
        return;
      }
      await runtime.storage.deleteReminder(reminderId);
      await runtime.scheduler.reconcile().catch((error) =>
        logError("reminder_reconcile_failed", error, {
          interaction_id: interaction.id,
          reminder_id: reminderId,
        }),
      );
      await finish(
        interaction,
        runtime,
        `Reminder ${reminderId} has been deleted successfully!`,
      );
    });
  },
};
